File: Common/src/main/java/ca/lukegrahamlandry/mercenaries/wrapped/GlobalMercListData.java
package ca.lukegrahamlandry.mercenaries.wrapped;

import ca.lukegrahamlandry.mercenaries.MercenariesMod;
import ca.lukegrahamlandry.mercenaries.entity.MercenaryEntity;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.UUID;
import java.util.function.Consumer;

public class GlobalMercListData {
    HashMap<UUID, ArrayList<UUID>> mercs;

    public void addMerc(ServerPlayer player, MercenaryEntity merc){
        ArrayList<UUID> owned = mercs.getOrDefault(player.getUUID(), new ArrayList<>());
        owned.add(merc.getUUID());
        mercs.put(player.getUUID(), owned);
        MercenariesMod.MERC_LIST.setDirty();
    }

    public void removeMerc(ServerPlayer player, MercenaryEntity merc){
        ArrayList<UUID> owned = mercs.getOrDefault(player.getUUID(), new ArrayList<>());
        owned.remove(merc.getUUID());
        MercenariesMod.MERC_LIST.setDirty();
    }

    public ArrayList<UUID> getMercs(ServerPlayer player){
        return mercs.getOrDefault(player.getUUID(), new ArrayList<>());
    }

    public void forLoadedMercBelongingTo(ServerPlayer player, Consumer<MercenaryEntity> action){
        for (UUID mID : getMercs(player)){
            Entity maybeMerc = ((ServerLevel)player.level).getEntity(mID);
            if (maybeMerc instanceof MercenaryEntity){
                action.accept((MercenaryEntity) maybeMerc);
            }
        }
    }
}

File: Common/src/main/kotlin/ca/lukegrahamlandry/mercenaries/data/GlobalMercListData.kt
package ca.lukegrahamlandry.mercenaries.data

import ca.lukegrahamlandry.mercenaries.MercenariesMod
import ca.lukegrahamlandry.mercenaries.entity.MercenaryEntity
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.player.Player
import java.util.*
import java.util.function.Consumer

data class GlobalMercListData(private var mercs: HashMap<UUID, ArrayList<UUID>> = HashMap<UUID, ArrayList<UUID>>()) {
    fun addMerc(player: Player, merc: MercenaryEntity) {
        val owned = mercs.getOrDefault(player.uuid, ArrayList())
        owned.add(merc.uuid)
        mercs[player.uuid] = owned
        MercenariesMod.MERC_LIST.setDirty()
    }

    fun removeMerc(player: UUID, merc: MercenaryEntity) {
        val owned = mercs.getOrDefault(player, ArrayList())
        owned.remove(merc.uuid)
        MercenariesMod.MERC_LIST.setDirty()
    }

    fun getMercs(player: Player): ArrayList<UUID> {
        return mercs.getOrDefault(player.uuid, ArrayList())
    }

    fun forLoadedMercBelongingTo(player: Player, action: Consumer<MercenaryEntity?>) {
        if (player.level.isClientSide()) throw RuntimeException("Cannot call GlobalMercListData#forLoadedMercBelongingTo on the client")
        for (mID in getMercs(player)) {
            val maybeMerc = (player.level as ServerLevel).getEntity(mID)
            if (maybeMerc is MercenaryEntity) {
                action.accept(maybeMerc as MercenaryEntity?)
            }
        }
    }
}